subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/DefaultGradleApiSourcesResolver.java

```diff
@@ -16,20 +16,128 @@
 
 package org.gradle.plugins.ide.internal.resolver;
 
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.repositories.IvyArtifactRepository;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration;
+import org.gradle.api.internal.artifacts.transform.UnzipTransform;
+import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
 import org.gradle.internal.installation.CurrentGradleInstallation;
 import org.gradle.internal.installation.GradleInstallation;
+import org.gradle.util.GFileUtils;
 
 import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
 
 public class DefaultGradleApiSourcesResolver implements GradleApiSourcesResolver {
 
+    private static final Logger LOGGER = Logging.getLogger(DefaultGradleApiSourcesResolver.class);
+
+    private static final String GRADLE_REPO_URL = "https://services.gradle.org/";
+    private static final String GRADLE_REPO_URL_OVERRIDE_VAR = "GRADLE_REPO_OVERRIDE";
+
+    private final ProjectInternal project;
+
+    private final Attribute<String> artifactType = Attribute.of("artifactType", String.class);
+    private final String zipType = "zip";
+    private final String unzippedDistributionType = "unzipped-distribution";
+    private final String sourceDirectory = "src-directory";
+
+    public DefaultGradleApiSourcesResolver(ProjectInternal project) {
+        this.project = project;
+    }
+
     @Override
-    public File resolveGradleApiSources(File artifact) {
+    public File resolveGradleApiSources(boolean download) {
         GradleInstallation gradleInstallation = CurrentGradleInstallation.get();
-        if (gradleInstallation == null || !gradleInstallation.getSrcDir().exists()) {
+        if (gradleInstallation == null) {
             return null;
         }
-        return gradleInstallation.getSrcDir();
+        File srcDir = gradleInstallation.getSrcDir();
+        if (!srcDir.exists()) {
+            if (!download) {
+                return null;
+            }
+            try {
+                File sources = downloadSources();
+                GFileUtils.moveExistingDirectory(sources, srcDir);
+            } catch (DefaultLenientConfiguration.ArtifactResolveException e) {
+                LOGGER.warn("Could not fetch Gradle sources distribution: " + e.getCause().getMessage());
+                return null;
+            }
+        }
+        return srcDir;
+    }
+
+    private File downloadSources() {
+        IvyArtifactRepository repository = addGradleSourcesRepository();
+        try {
+            registerTransforms();
+            return transientConfigurationForSourcesDownload();
+        } finally {
+            project.getRepositories().remove(repository);
```

Comment:
This repository is added just above the try block and is needed only to resolve this configuration. This resolver runs only for `eclipse` and `idea` plugin tasks - does it still mean it can be intercepted by a user?

This is also how KotlinDSL implementation does it: https://github.com/gradle/gradle/blob/master/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/SourceDistributionProvider.kt#L63-L68

Reply:
I think it is ok like this. As Vaidotas pointed out it's in line with what we already do in SourceDistributionProvider.kt.

In theory, user code can intercept the repository add/remove. But we need at least the add in any case. I think it is good to have the remove as well then. This code will usually only be executed in IDE Sync etc. via Tooling API. If you run any of you normal build tasks, this won't affect the build.

It's maybe not perfect, but I also don't see a more elegant solution with what we have at the moment.